traversal/iterator/GraphIterator.java

```diff
@@ -360,48 +354,55 @@ public Scoped get(Identifier.Variable scope) {
 
         public static class Scoped {
 
-            private final Set<ThingVertex> roleInstances;
-            private final PriorityQueue<Integer> edgeOrders;
+            Set<ThingVertex> roles;
+            TreeMap<Integer, ThingVertex> visited;
 
             private Scoped() {
-                this.roleInstances = new HashSet<>();
-                this.edgeOrders = new PriorityQueue<>(Comparator.reverseOrder());
+                this.roles = new HashSet<>();
+                visited = new TreeMap<>();
             }
 
             public Collection<Integer> edgeOrders() {
-                return edgeOrders;
+                return visited.keySet();
             }
 
             public boolean isEmpty() {
-                assert edgeOrders.isEmpty() == roleInstances.isEmpty();
-                return edgeOrders.isEmpty();
+                assert roles.size() == visited.size();
+                return visited.isEmpty();
             }
 
-            public Integer peekLastEdgeOrder() {
+            public Integer lastEdgeOrder() {
                 assert !isEmpty();
-                return edgeOrders.peek();
+                return visited.lastKey();
             }
 
             public boolean contains(ThingVertex roleVertex) {
-                return roleInstances.contains(roleVertex);
+                return roles.contains(roleVertex);
+            }
+
+            public void popLast() {
+                ThingVertex poppedRole = visited.remove(lastEdgeOrder());
+                assert roles.contains(poppedRole);
+                roles.remove(poppedRole);
             }
 
-            public void record(ThingVertex roleVertex, int order) {
-                assert !roleInstances.contains(roleVertex) && !edgeOrders.contains(order);
-                roleInstances.add(roleVertex);
-                edgeOrders.add(order);
+            public void push(ThingVertex role, int order) {
+                assert (isEmpty() || order > lastEdgeOrder()) && !roles.contains(role);
+                visited.put(order, role);
+                roles.add(role);
             }
 
-            public void remove(ThingVertex previousRole, int order) {
-                assert roleInstances.contains(previousRole) && edgeOrders.contains(order);
-                roleInstances.remove(previousRole);
-                edgeOrders.remove(order);
+            public void replace(int order, ThingVertex newRole) {
```

Comment:
i can't think of a fast way to "swap" out a role for the same order, all operations are essentially remove an element from a map and replace it. However, since we also have to retrieve the old value to delete it from the fast lookup set, we might as well just delete it